### PR TITLE
Improve Motion Constraints and Setup for Kickoffs

### LINKED ISSUE #2236: Improve motion constraints and setup for kickoffs
### Description of the task

During Robocup, we quickly created a new Kickoff play that allowed us to keep possession. In the 1-robocup branch, we made  `KickoffChipTactic` and `KickoffMoveTactic` which are both just copies of `KickTactic` and `MoveTactic` without any new functionality. We needed it because we needed to set special motion constraints in `motion_constraint_visitor.cpp` that let the kicker into the enemy half centre circle so that the ball can be kicked. We want a macro (possibly defined in `tactic.h`) that lets us duplicate/extend an existing tactic with a new name without needing to make new header and source files.

For this ticket, just make this macro and use it to make the current (the one in master) `KickoffChipTactic`, so that we can delete `kickoff_chip_tactic.h` and `kickoff_chip_tactic.cpp`.

Also, in `motion_constraint_set_builder.cpp`, adjust the conditions so that the motion constraint HALF_METER_AROUND_BALL is only active during kickoffs if we are still in the setup state.

### Acceptance criteria

- [ ] Make a macro that lets us duplicate/extend tactics with new names
- [ ] Use this macro to make `KickoffChipTactic` and delete `kickoff_chip_tactic.h` and `kickoff_chip_tactic.cpp`
- [ ] Adjust the HALF_METER_AROUND_BALL motion constraint for kickoffs
- [ ] Make sure that motion constraint behaviour looks correct by looking at some kickoff gameplay on AI vs AI

### PR DESCRIPTION
**Pull request by BrianS99**
### Description
Deleted `kickoff_chip_tactic.cpp` and `kickoff_chip_tactic.h` and added a macro in `tactic.h` that allows us to duplicate tactic classes. Recreated KickoffChipTactic class using ChipTactic as a copy. Also changed `motion_constraint_set_builder.cpp` so that HALF_METER_AROUND_BALL is only active during kickoffs if we are still in the setup state.

### Testing Done
Reviewed AI vs. AI plays on simulator and it looks normal.

### Resolved Issues
Resolves #2236 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/ai/hl/stp/tactic/tactic.h [hunk 1]
```diff
@@ -10,6 +10,12 @@
 #include "software/ai/intent/intent.h"
 #include "software/world/world.h"
 
+#define COPY_CLASS(new_class, parent_class)                                              \
+    class new_class : public parent_class                                                \
+    {                                                                                    \
+        using parent_class::parent_class;                                                \
+    };
+
```
Comment (jonathanlew) [inline src/software/ai/hl/stp/tactic/tactic.h]:
```suggestion
#define COPY_CLASS(new_class, parent_class)                                              \
    class new_class : public parent_class                                                \
    {                                                                                    \
        using parent_class::parent_class;                                                \
                                                                                         \
        void accept(TacticVisitor &visitor) const                                        \
        {                                                                                \
            visitor.visit(*this);                                                        \
        }                                                                                \
    };

```
The tactic visitor needs to be overridden to have different motion constraints

File: src/software/ai/motion_constraint/motion_constraint_set_builder.cpp [hunk 2]
```diff
@@ -56,9 +56,12 @@ std::set<MotionConstraint> buildMotionConstraintSetFromGameState(
     }
     else if (game_state.isKickoff())
     {
-        motion_constraints.insert({MotionConstraint::HALF_METER_AROUND_BALL,
-                                   MotionConstraint::CENTER_CIRCLE,
-                                   MotionConstraint::ENEMY_HALF});
+        motion_constraints.insert(
```
Comment (jonathanlew) [inline src/software/ai/motion_constraint/motion_constraint_set_builder.cpp:59]: I'm not sure why these changes are needed

**Review by jonathanlew — CHANGES_REQUESTED** [review]
Good start, I found an issue with the macro that needs fixing

File: src/software/ai/motion_constraint/motion_constraint_set_builder.cpp (hunk 2 above)
Comment (BrianS99) [reply src/software/ai/motion_constraint/motion_constraint_set_builder.cpp:59]: This was in the TODO in the ticket, we should ask Arun.

Comment (jonathanlew) [reply src/software/ai/motion_constraint/motion_constraint_set_builder.cpp:59]: I see, ignore my comment then

File: src/software/ai/hl/stp/play/BUILD [hunk 3]
```diff
@@ -189,7 +189,7 @@ cc_library(
         "//shared:constants",
         "//shared/parameter:cpp_configs",
         "//software/ai/evaluation:enemy_threat",
-        "//software/ai/hl/stp/tactic:kickoff_chip_tactic",
+        "//software/ai/hl/stp/tactic:all_tactics",
```
Comment (jonathanlew) [inline src/software/ai/hl/stp/play/BUILD]: I don't think this dependency is necessary

**Review by jonathanlew — COMMENTED** [review]
just one nit, otherwise LGTM!

File: src/software/ai/hl/stp/tactic/chip/chip_tactic.h [hunk 4]
```diff
@@ -69,3 +69,5 @@ class ChipTactic : public Tactic
     // Tactic parameters
     ChipFSM::ControlParams control_params;
 };
+
```
Comment (akhilveeraghanta) [inline src/software/ai/hl/stp/tactic/chip/chip_tactic.h:72]: nit: plz add a comment here explaining what this does

File: src/software/ai/hl/stp/tactic/move/move_tactic.h [hunk 5]
```diff
@@ -81,3 +81,5 @@ class MoveTactic : public Tactic
 
     MoveFSM::ControlParams control_params;
 };
+
```
Comment (akhilveeraghanta) [inline src/software/ai/hl/stp/tactic/move/move_tactic.h:84]: ditto above

File: src/software/ai/hl/stp/tactic/tactic.h [hunk 6]
```diff
@@ -10,6 +10,17 @@
 #include "software/ai/intent/intent.h"
 #include "software/world/world.h"
```
Comment (akhilveeraghanta) [inline src/software/ai/hl/stp/tactic/tactic.h:12]: nit: missing javadocs

**Review by akhilveeraghanta — CHANGES_REQUESTED** [review]
very cool, just a couple comments

File: src/software/ai/hl/stp/tactic/tactic.h (hunk 6 above)
Comment (BrianS99) [reply src/software/ai/hl/stp/tactic/tactic.h:12]: Fixed

File: src/software/ai/hl/stp/tactic/move/move_tactic.h (hunk 5 above)
Comment (BrianS99) [reply src/software/ai/hl/stp/tactic/move/move_tactic.h:84]: Fixed

File: src/software/ai/hl/stp/tactic/chip/chip_tactic.h (hunk 4 above)
Comment (BrianS99) [reply src/software/ai/hl/stp/tactic/chip/chip_tactic.h:72]: Fixed

File: src/software/ai/hl/stp/play/BUILD (hunk 3 above)
Comment (BrianS99) [reply src/software/ai/hl/stp/play/BUILD]: Fixed

**Review by akhilveeraghanta — DISMISSED** [review]
lgtm!

**Review by jonathanlew — DISMISSED** [review]
LGTM! (pending CI)

**Review by jameschuang2002 — DISMISSED** [review]
lgtm

**Review by jonathanlew — APPROVED** [review]
LGTM!

**Review by akhilveeraghanta — APPROVED** [review]
